File: golang/group_e2ee/v2_aad.go
package groupe2ee

import (
	"bytes"
	"encoding/json"

	"github.com/agent-network-protocol/anp/golang/internal/cjson"
)

func GroupSendAuthenticatedDataV2(meta V2GroupSendMetadata, body V2GroupCipherObject) ([]byte, error) {
	if err := meta.Validate(); err != nil {
		return nil, err
	}
	if err := body.Validate(); err != nil {
		return nil, err
	}
	if body.GroupStateRef.GroupDID != meta.Target.DID {
		return nil, invalidV2("body.group_state_ref.group_did must equal meta.target.did")
	}
	return cjson.Marshal(map[string]any{
		"content_type":         GroupCipherContentTypeV2,
		"group_did":            meta.Target.DID,
		"crypto_group_id_b64u": body.CryptoGroupIDB64U,
		"group_state_ref":      body.GroupStateRef,
		"security_profile":     SecurityProfileV2,
		"sender_did":           meta.SenderDID,
		"sender_device_id":     meta.SenderDeviceID,
		"message_id":           meta.MessageID,
		"operation_id":         meta.OperationID,
	})
}

func groupMembershipSubmissionBindingV2(method string, meta V2GroupControlMetadata, body any) ([]byte, error) {
	if method != MethodGroupAddV2 && method != MethodGroupRemoveV2 {
		return nil, invalidV2("subject_method must be group.e2ee.add or group.e2ee.remove")
	}
	if err := meta.Validate(); err != nil {
		return nil, err
	}
	var memberDID, memberDeviceID, cryptoGroupID, epoch string
	var stateRef V2GroupStateRef
	switch typed := body.(type) {
	case V2GroupAddBody:
		if method != MethodGroupAddV2 {
			return nil, invalidV2("add body requires group.e2ee.add")
		}
		if err := typed.Validate(); err != nil {
			return nil, err
		}
		memberDID, memberDeviceID, stateRef, cryptoGroupID, epoch = typed.MemberDID, typed.MemberDeviceID, typed.GroupStateRef, typed.CryptoGroupIDB64U, typed.Epoch
	case V2GroupRemoveBody:
		if method != MethodGroupRemoveV2 {
			return nil, invalidV2("remove body requires group.e2ee.remove")
		}
		if err := typed.Validate(); err != nil {
			return nil, err
		}
		memberDID, memberDeviceID, stateRef, cryptoGroupID, epoch = typed.MemberDID, typed.MemberDeviceID, typed.GroupStateRef, typed.CryptoGroupIDB64U, typed.Epoch
	default:
		return nil, invalidV2("unsupported membership body")
	}
	if stateRef.GroupDID != meta.Target.DID {
		return nil, invalidV2("group_state_ref.group_did must equal meta.target.did")
	}
	return cjson.Marshal(map[string]any{
		"group_did":            meta.Target.DID,
		"crypto_group_id_b64u": cryptoGroupID,
		"group_state_ref":      stateRef,
		"subject_method":       method,
		"member_did":           memberDID,
		"member_device_id":     memberDeviceID,
		"epoch":                epoch,
		"security_profile":     SecurityProfileV2,
		"sender_did":           meta.SenderDID,
		"sender_device_id":     meta.SenderDeviceID,
		"operation_id":         meta.OperationID,
	})
}

func GroupAddSubmissionBindingV2(meta V2GroupControlMetadata, body V2GroupAddBody) ([]byte, error) {
	return groupMembershipSubmissionBindingV2(MethodGroupAddV2, meta, body)
}

func GroupRemoveSubmissionBindingV2(meta V2GroupControlMetadata, body V2GroupRemoveBody) ([]byte, error) {
	return groupMembershipSubmissionBindingV2(MethodGroupRemoveV2, meta, body)
}

func CanonicalGroupApplicationPlaintextV2(value V2GroupApplicationPlaintext) ([]byte, error) {
	if err := value.Validate(); err != nil {
		return nil, err
	}
	return cjson.Marshal(value)
}

func ParseGroupApplicationPlaintextV2(value any) (V2GroupApplicationPlaintext, error) {
	encoded, err := json.Marshal(value)
	if err != nil {
		return V2GroupApplicationPlaintext{}, err
	}
	var members map[string]json.RawMessage
	if err := json.Unmarshal(encoded, &members); err != nil {
		return V2GroupApplicationPlaintext{}, invalidV2("group application plaintext must be an object")
	}
	for _, field := range []string{"thread_id", "reply_to_message_id", "annotations", "text", "payload", "payload_b64u"} {
		if raw, present := members[field]; present && bytes.Equal(bytes.TrimSpace(raw), []byte("null")) {
			return V2GroupApplicationPlaintext{}, invalidV2(field + " must be omitted rather than null")
		}
	}
	var plaintext V2GroupApplicationPlaintext
	if err := decodeStrictV2(encoded, &plaintext); err != nil {
		return V2GroupApplicationPlaintext{}, err
	}
	if err := plaintext.Validate(); err != nil {
		return V2GroupApplicationPlaintext{}, err
	}
	return plaintext, nil
}

File: rust/src/group_e2ee/v2_aad.rs
use serde_json::{json, Value};

use crate::canonical_json::canonicalize_json;

use super::v2_errors::GroupE2eeV2Error;
use super::v2_models::{
    V2GroupAddBody, V2GroupApplicationPlaintext, V2GroupRemoveBody, V2GroupSendMetadata,
    GROUP_CIPHER_CONTENT_TYPE_V2, GROUP_E2EE_SECURITY_PROFILE_V2, METHOD_GROUP_ADD_V2,
    METHOD_GROUP_REMOVE_V2,
};

/// Build the exact RFC 8785 bytes carried as MLS `authenticated_data` by P6 v2 send.
pub fn group_send_authenticated_data_v2(
    meta: &V2GroupSendMetadata,
    body: &super::v2_models::V2GroupCipherObject,
) -> Result<Vec<u8>, GroupE2eeV2Error> {
    meta.validate()?;
    body.validate()?;
    if body.group_state_ref.group_did != meta.target.did {
        return Err(GroupE2eeV2Error::invalid(
            "body.group_state_ref.group_did must equal meta.target.did",
        ));
    }
    canonicalize_json(&json!({
        "content_type": GROUP_CIPHER_CONTENT_TYPE_V2,
        "group_did": meta.target.did,
        "crypto_group_id_b64u": body.crypto_group_id_b64u,
        "group_state_ref": body.group_state_ref,
        "security_profile": GROUP_E2EE_SECURITY_PROFILE_V2,
        "sender_did": meta.sender_did,
        "sender_device_id": meta.sender_device_id,
        "message_id": meta.message_id,
        "operation_id": meta.operation_id,
    }))
    .map_err(GroupE2eeV2Error::from)
}

trait V2MembershipBody {
    fn member_did(&self) -> &str;
    fn member_device_id(&self) -> &str;
    fn group_state_ref(&self) -> &super::v2_models::V2GroupStateRef;
    fn crypto_group_id_b64u(&self) -> &str;
    fn epoch(&self) -> &str;
    fn validate_membership(&self) -> Result<(), GroupE2eeV2Error>;
}

impl V2MembershipBody for V2GroupAddBody {
    fn member_did(&self) -> &str {
        &self.member_did
    }
    fn member_device_id(&self) -> &str {
        &self.member_device_id
    }
    fn group_state_ref(&self) -> &super::v2_models::V2GroupStateRef {
        &self.group_state_ref
    }
    fn crypto_group_id_b64u(&self) -> &str {
        &self.crypto_group_id_b64u
    }
    fn epoch(&self) -> &str {
        &self.epoch
    }
    fn validate_membership(&self) -> Result<(), GroupE2eeV2Error> {
        self.validate()
    }
}

impl V2MembershipBody for V2GroupRemoveBody {
    fn member_did(&self) -> &str {
        &self.member_did
    }
    fn member_device_id(&self) -> &str {
        &self.member_device_id
    }
    fn group_state_ref(&self) -> &super::v2_models::V2GroupStateRef {
        &self.group_state_ref
    }
    fn crypto_group_id_b64u(&self) -> &str {
        &self.crypto_group_id_b64u
    }
    fn epoch(&self) -> &str {
        &self.epoch
    }
    fn validate_membership(&self) -> Result<(), GroupE2eeV2Error> {
        self.validate()
    }
}

fn membership_submission_binding<B: V2MembershipBody>(
    method: &str,
    meta: &super::v2_models::V2GroupControlMetadata,
    body: &B,
) -> Result<Vec<u8>, GroupE2eeV2Error> {
    if !matches!(method, METHOD_GROUP_ADD_V2 | METHOD_GROUP_REMOVE_V2) {
        return Err(GroupE2eeV2Error::invalid(
            "subject_method must be group.e2ee.add or group.e2ee.remove",
        ));
    }
    meta.validate()?;
    body.validate_membership()?;
    if body.group_state_ref().group_did != meta.target.did {
        return Err(GroupE2eeV2Error::invalid(
            "group_state_ref.group_did must equal meta.target.did",
        ));
    }
    canonicalize_json(&json!({
        "group_did": meta.target.did,
        "crypto_group_id_b64u": body.crypto_group_id_b64u(),
        "group_state_ref": body.group_state_ref(),
        "subject_method": method,
        "member_did": body.member_did(),
        "member_device_id": body.member_device_id(),
        "epoch": body.epoch(),
        "security_profile": GROUP_E2EE_SECURITY_PROFILE_V2,
        "sender_did": meta.sender_did,
        "sender_device_id": meta.sender_device_id,
        "operation_id": meta.operation_id,
    }))
    .map_err(GroupE2eeV2Error::from)
}

/// Build the exact P6 v2 Add submission binding without null/default placeholders.
pub fn group_add_submission_binding_v2(
    meta: &super::v2_models::V2GroupControlMetadata,
    body: &V2GroupAddBody,
) -> Result<Vec<u8>, GroupE2eeV2Error> {
    membership_submission_binding(METHOD_GROUP_ADD_V2, meta, body)
}

/// Build the exact P6 v2 Remove submission binding without null/default placeholders.
pub fn group_remove_submission_binding_v2(
    meta: &super::v2_models::V2GroupControlMetadata,
    body: &V2GroupRemoveBody,
) -> Result<Vec<u8>, GroupE2eeV2Error> {
    membership_submission_binding(METHOD_GROUP_REMOVE_V2, meta, body)
}

/// Canonicalize the inner group application object before MLS encryption.
pub fn canonical_group_application_plaintext_v2(
    plaintext: &V2GroupApplicationPlaintext,
) -> Result<Vec<u8>, GroupE2eeV2Error> {
    plaintext.validate()?;
    let value: Value = serde_json::to_value(plaintext)?;
    canonicalize_json(&value).map_err(GroupE2eeV2Error::from)
}

pub fn parse_group_application_plaintext_v2(
    value: &Value,
) -> Result<V2GroupApplicationPlaintext, GroupE2eeV2Error> {
    let plaintext: V2GroupApplicationPlaintext = serde_json::from_value(value.clone())?;
    plaintext.validate()?;
    Ok(plaintext)
}
